### ISSUE TITLE
twisted.words.protocols.irc: Don't prepend # onto channel name in IRCClient.me

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3910]
|<img alt="stefanor's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| stefanor reported|
|-|-|
|Trac ID|trac#3910|
|Type|defect|
|Created|2009-07-11 11:55:58Z|

You can't send an action privately (not in a channel) using ```IRCClient```, because ```IRCClient.me``` prepends a ```#``` onto the channel name. While that behaviour is documented in docstrings elsewhere (See ticket [#3895](https://github.com/twisted/twisted/issues/3895)), it is a bad API choice.

Without this patch, the alternative is to call ```IRCClient.ctcpMakeQuery``` directly.

Attachments:

* [t.w.p.irc-action.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dd3/dd3a575117170389f7a4185aac816f1c72068754/c8fe61e8e680e6ad048bc1b715843032ce879344.patch) (2551 bytes) - added by stefanor on 2009-07-11 18:16:12Z - describe() function replaces me() and doesn't prepend #
* [t.w.p.irc-action-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dd3/dd3a575117170389f7a4185aac816f1c72068754/64f3adecb78d32abd59441025d145ff93d1d5fc0.patch) (3739 bytes) - added by stefanor on 2009-07-12 18:02:48Z - Improve tests and test deprecation

<details><summary>Searchable metadata</summary>

```
trac-id__3910 3910
type__defect defect
reporter__stefanor stefanor
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1247313358000000 1247313358000000
changetime__1247497633000000 1247497633000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to stefanor|
|-|-|

This is what the API does.  Changing it may break existing software which relies on it.

Instead, a new method with the desired behavior should be introduced, and possibly the existing method should be deprecated.

Please be sure to include unit tests along with any change. :)

**Comment by twisted-trac** [comment]
|<img alt="stefanor's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|stefanor commented|
|-|-|

Replying to [exarkun](#note_1):
> Instead, a new method with the desired behavior should be introduced, and possibly the existing method should be deprecated.

All the good names are taken. the best I can come up with is ```describe()```

> Please be sure to include unit tests along with any change. :)

Bleh, OK.

**Comment by twisted-trac** [comment]
|<img alt="stefanor's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stefanor set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to stefanor|
|-|-|

Thanks for the new patch.  I just have a few comments:

  1. The new test method, `test_describe`, should have a docstring describing what behavior it covers.  You can see examples of such docstrings just above it on the other test methods in that class.
  1. The Twisted coding standard calls for two blank lines between method definitions; please add a blank line before `test_describe` and after `describe`.
  1. The `me` deprecation should also be unit tested.  Unfortunately this is pretty old code, so `me` doesn't have _any_ existing unit tests.  Also unfortunately, Twisted hasn't quite settled on a convention for testing deprecations, so there are a few possibilities for you to choose from here. :)  One style involves `TestCase.flushWarnings`, you can find an example of this style in [test_util.py](http://twistedmatrix.com/trac/browser/trunk/twisted/python/test/test_util.py?rev=26967#L118).  Another style involves `TestCase.callDeprecated`, and you can find an example in [test_keys.py](http://twistedmatrix.com/trac/browser/trunk/twisted/conch/test/test_keys.py?rev=27079#L424) (this involves a different implementation of emitting the warning too, see [asn1.py](http://twistedmatrix.com/trac/browser/trunk/twisted/conch/ssh/asn1.py?rev=27079).  There's also `TestCase.assertWarns`, but either of the previous forms is probably preferable to that.
  1. The deprecation should mention that the API was first deprecated in Twisted 9.0.  Depending on how you resolve the previous point, this may be fixed automatically.
  1. The new `describe` method should include the epytext markup `@since: 9.0` in its docstring.

That's all.  Thanks again for working on this.

**Comment by twisted-trac** [comment]
|<img alt="stefanor's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|stefanor set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_4):
> Thanks again for working on this.

OK, I admit I was lazier in the earlier patch attempts :P

Hope this solves all the issues.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Thanks, looks great.  I'll apply it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [27131]) Apply t.w.p.irc-action-2.patch adding a better API than IRCClient.me

Author: stefanor
Reviewer: exarkun
Fixes: [#3910](https://github.com/twisted/twisted/issues/3910)

Add `IRCClient.describe`, a replacement for `IRCClient.me` which doesn't do
confusing, unhelpful target mangling.  Also deprecate `IRCClient.me`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
